crates/nu-command/src/filesystem/cp.rs

```diff
@@ -345,29 +377,99 @@ fn interactive_copy(
     }
 }
 
+// `copy_file` uses `std::fs::copy` to copy a file. There is another function called `copy_file_with_progressbar`
+// which uses `read` and `write` instead. This is to get the progress of the copy. If you make any changes in this
+// function try to match the changes in `copy_file_with_progressbar`
 fn copy_file(src: PathBuf, dst: PathBuf, span: Span) -> Value {
     match std::fs::copy(&src, &dst) {
         Ok(_) => {
             let msg = format!("copied {:} to {:}", src.display(), dst.display());
             Value::String { val: msg, span }
         }
-        Err(e) => {
-            let message = format!("copy file {src:?} failed: {e}");
-
-            use std::io::ErrorKind;
-            let shell_error = match e.kind() {
-                ErrorKind::NotFound => ShellError::FileNotFoundCustom(message, span),
-                ErrorKind::PermissionDenied => ShellError::PermissionDeniedError(message, span),
-                ErrorKind::Interrupted => ShellError::IOInterrupted(message, span),
-                ErrorKind::OutOfMemory => ShellError::OutOfMemoryError(message, span),
-                // TODO: handle ExecutableFileBusy etc. when io_error_more is stabilized
-                // https://github.com/rust-lang/rust/issues/86442
-                _ => ShellError::IOErrorSpanned(message, span),
-            };
-
-            Value::Error { error: shell_error }
-        }
+        Err(e) => error_return(e, src, span),
+    }
+}
+
+// `copy_file_with_progressbar` uses `read` and `write` to copy a file. There is another function called `copy_file`
+// which uses `std::fs::copy` instead which is faster but can't gt the process of the copy. If you make any changes
+// in this function try to match the changes in `copy_file`
+fn copy_file_with_progressbar(src: PathBuf, dst: PathBuf, span: Span) -> Value {
+    let mut bytes_processed: u64 = 0;
+    let bytes_processed_p = &mut bytes_processed;
+    let mut process_failed: Option<std::io::Error> = None;
+    let process_failed_p = &mut process_failed;
```

Comment:
What are the `_p` variables for? I don't think they're needed.

Reply:
those are pointers

Reply:
Gotcha. You shouldn't need variables for those, you should be able to get by with this:

```suggestion
    let mut bytes_processed: u64 = 0;
    let mut process_failed: Option<std::io::Error> = None;
```